kafka/signed_token_redeem_handler.go

```diff
@@ -62,36 +94,57 @@ func SignedTokenRedeemHandler(
 			redeemedTokenResults = append(redeemedTokenResults, avroSchema.RedeemResult{
 				Issuer_name:     "",
 				Issuer_cohort:   0,
-				Status:          redeemError,
+				Status:          avroSchema.RedeemResultStatusError,
 				Associated_data: request.Associated_data,
 			})
 			continue
 		}
 
+		// preimage, signature, and binding are all required to proceed
 		if request.Token_preimage == "" || request.Signature == "" || request.Binding == "" {
 			logger.Error().
 				Err(fmt.Errorf("request %s: empty request", tokenRedeemRequestSet.Request_id)).
 				Msg("signed token redeem handler")
 			redeemedTokenResults = append(redeemedTokenResults, avroSchema.RedeemResult{
 				Issuer_name:     "",
 				Issuer_cohort:   0,
-				Status:          redeemError,
+				Status:          avroSchema.RedeemResultStatusError,
 				Associated_data: request.Associated_data,
 			})
 			continue
 		}
 
 		tokenPreimage := crypto.TokenPreimage{}
 		err = tokenPreimage.UnmarshalText([]byte(request.Token_preimage))
+		// Unmarshaling failure is a data issue and is probably permanent.
 		if err != nil {
-			return fmt.Errorf("request %s: could not unmarshal text into preimage: %w",
-				tokenRedeemRequestSet.Request_id, err)
+			message := fmt.Sprintf("request %s: could not unmarshal text into preimage", tokenRedeemRequestSet.Request_id)
+			processingResult, errorResult := avroRedeemErrorResultFromError(
```

Comment:
non-blocking nit: it seems like every time we call `avroRedeemErrorResultFromError(...)`, we immediately call `MayEmitIfPermanent(...)` and then `return NilIfPermanent(...)`. it seems like this could all be a single `handleRedeemError` or similar?

Reply:
Worth the effort. Good call.